isSubTypeOf on TypeReference is not robust against missing declarations

`a.isSubTypeOf(b)` from `TypeReference` is not robust and throws a `NullPointerException` if the declaration of `a` is not available.

```
Caused by: java.lang.NullPointerException
	at spoon.support.reflect.reference.CtTypeParameterReferenceImpl.isSubtypeOf(CtTypeParameterReferenceImpl.java:278)
	at at.jku.isse.gradient.java.AsgRelationshipTransformer$Transformer.createElementType(AsgRelationshipTransformer.kt:393)
	at at.jku.isse.gradient.java.AsgRelationshipTransformer$Transformer.visitCtFieldReference(AsgRelationshipTransformer.kt:119)
	at spoon.support.reflect.reference.CtFieldReferenceImpl.accept(CtFieldReferenceImpl.java:58)
	at at.jku.isse.gradient.java.SpoonVisitor.scan(SpoonVisitor.kt:16)
	at at.jku.isse.gradient.java.AsgRelationshipTransformer$Transformer.scan(AsgRelationshipTransformer.kt:62)
	at at.jku.isse.gradient.java.SpoonVisitor.scan(SpoonVisitor.kt:10)
	at at.jku.isse.gradient.java.AsgRelationshipTransformer$Transformer$visitCtTypeReference$$inlined$let$lambda$1.accept(AsgRelationshipTransformer.kt:83)
	at at.jku.isse.gradient.java.AsgRelationshipTransformer$Transformer$visitCtTypeReference$$inlined$let$lambda$1.accept(AsgRelationshipTransformer.kt:26)
	at java.util.Optional.ifPresent(Optional.java:159)
	at at.jku.isse.gradient.java.AsgRelationshipTransformer$Transformer.visitCtTypeReference(AsgRelationshipTransformer.kt:73)
	at spoon.support.reflect.reference.CtTypeReferenceImpl.accept(CtTypeReferenceImpl.java:78)
	at at.jku.isse.gradient.java.SpoonVisitor.scan(SpoonVisitor.kt:16)
	at at.jku.isse.gradient.java.AsgRelationshipTransformer$Transformer.scan(AsgRelationshipTransformer.kt:62)
	at at.jku.isse.gradient.java.AsgRelationshipTransformer.transform(AsgRelationshipTransformer.kt:20)
	at at.jku.isse.gradient.java.ModelTransformer$transform$7.accept(ModelTransformer.kt:38)
	at at.jku.isse.gradient.java.ModelTransformer$transform$7.accept(ModelTransformer.kt:11)
	at java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:184)
	at java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1382)
	at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:481)
	at java.util.stream.ForEachOps$ForEachTask.compute(ForEachOps.java:291)
	at java.util.concurrent.CountedCompleter.exec(CountedCompleter.java:731)
	at java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:289)
	at java.util.concurrent.ForkJoinPool$WorkQueue.execLocalTasks(ForkJoinPool.java:1040)
	at java.util.concurrent.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1058)
	at java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1692)
	at java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:157)
```

**Versions**
Spoon 7.0.0 (but also in early versions)

--- a/src/test/java/spoon/test/ctType/CtTypeTest.java
+++ b/src/test/java/spoon/test/ctType/CtTypeTest.java
@@ -163,7 +163,7 @@ public class CtTypeTest {
 
 		assertTrue(count[0] > (9 * 8));
 
-		// contract: isSubTypeOf does not throw an exception
+		// contract: isSubTypeOf does not throw any exception
 		// #2288 cannot be reproduced, probably fixed by #2406
 		CtTypeReference<Object> typeReferenceWithNoDeclaration = launcher.getFactory().createTypeReference();
 		typeReferenceWithNoDeclaration.setSimpleName("DoesNotExist");

--- a/src/test/java/spoon/test/ctType/CtTypeTest.java
+++ b/src/test/java/spoon/test/ctType/CtTypeTest.java
@@ -162,6 +162,13 @@ public class CtTypeTest {
 		});
 
 		assertTrue(count[0] > (9 * 8));
+
+		// contract: isSubTypeOf does not throw an exception
+		// #2288 cannot be reproduced, probably fixed by #2406
+		CtTypeReference<Object> typeReferenceWithNoDeclaration = launcher.getFactory().createTypeReference();
+		typeReferenceWithNoDeclaration.setSimpleName("DoesNotExist");
+		assertFalse(typeReferenceWithNoDeclaration.isSubtypeOf(oCtType.getReference()));
+		assertFalse(oCtType.isSubtypeOf(typeReferenceWithNoDeclaration));
 	}
 
 	private void checkIsSubtype(CtTypeReference superType, CtTypeReference subType, Map<String, CtTypeReference<?>> nameToTypeRef) {

